File: python/src/queries/accuracy_check_query.py
import urllib.request
import json
import os
import re
from typing import Any
from list_handler import __csv_to_dicts__

INVALID_PATTERNS = r'[\\\'\|\`\^\"\<\>\)\(\}\{\]\[\;\/\?\:\@\&\=\+\$\,\%\# ]'

class AccuracyCheckQuery:
    def __init__(self, scheme: str, host: str, bot_id: str, user_id: str, test_data: str) -> None:
        self.scheme    = re.sub(INVALID_PATTERNS, '', scheme)
        self.host      = re.sub(INVALID_PATTERNS, '', host)
        self.bot_id    = re.sub(INVALID_PATTERNS, '', bot_id)
        self.user_id   = re.sub(INVALID_PATTERNS, '', user_id)
        self.test_data = __csv_to_dicts__(test_data)
        self.uri       = f'{self.scheme}://{self.host}/api/v1/bots/{self.bot_id}/converse/{self.user_id}/secured?include=suggestions'

    def res_bodies(self) -> list[dict[str, Any]]:
        return [json.loads(res_body.read()) for res_body in self.__request__()]

    # private

    def __request__(self) -> list[Any]:
        responses = list()
        for test_datum in self.test_data:
            header = {
                'Content-Type': 'application/json',
                'Authorization': os.environ['BOTPRESS_ACCESS_TOKEN']
            }
            body = {
                'type': 'text',
                'text': test_datum['Question']
            }
            req = urllib.request.Request(self.uri, json.dumps(body).encode(), header)
            res = urllib.request.urlopen(req)
            responses.append(res)
        return responses

File: ruby/src/queries/accuracy_check_query.rb
require 'net/http'
require 'csv'
require 'json'

module Queries
  class AccuracyCheckQuery

    INVALID_PATTERNS = /[\\\'\|\`\^\"\<\>\)\(\}\{\]\[\;\/\?\:\@\&\=\+\$\,\%\# ]/

    def initialize(scheme, host, bot_id, user_id, test_data)
      @scheme    = scheme.gsub(INVALID_PATTERNS, '')
      @host      = host.gsub(INVALID_PATTERNS, '')
      @bot_id    = bot_id.gsub(INVALID_PATTERNS, '')
      @user_id   = user_id.gsub(INVALID_PATTERNS, '')
      @test_data = CSV.read(test_data, headers: true)
    end

    def res_bodies
      authorized
      request.map(&:body).map do |res_body|
        JSON.parse(res_body)
      end
    end

    private

    attr_reader :scheme, :host, :bot_id, :user_id, :test_data, :req

    def uri
      URI.parse("#{scheme}://#{host}/api/v1/bots/#{bot_id}/converse/#{user_id}/secured?include=suggestions")
    end

    def authorized
      @req                 = Net::HTTP::Post.new(uri)
      @req[:authorization] = ENV['BOTPRESS_ACCESS_TOKEN']
    end

    def request
      test_data['Question'].map { |question|
        req.set_form_data(type: :text, text: question)
        net_http         = Net::HTTP.new(uri.host, uri.port)
        net_http.use_ssl = true if uri.to_s.include?('https')
        net_http.start { it.request(req) }
      }
    end
  end
end
